### PR TITLE
[Merged by Bors] - feat: Mason-Stothers theorem (polynomial ABC)

### PR DESCRIPTION
**Pull request by seewoo5**
This file proves Mason-Stothers theorem (i.e. polynomial ABC). For coprime polynomials $a, b, c$ over a field (of arbitrary characteristic) with $a + b + c = 0$, we have $\max\{\deg(a), \deg(b), \deg(c)\} + 1 \le \deg(\mathrm{rad}(abc))$ or $a' = b' = c' = 0$.

---

- [x] depends on: #14873
- [x] depends on: #15331
- [x] depends on: #18452
- [x] depends on: #18483

Almost last part of porting project of the [formalization of polynomial ABC](https://github.com/seewoo5/lean-poly-abc).

TODO: After #14720 is merged, add polynomial FLT as a corollary. It is already in [here](https://github.com/seewoo5/lean-poly-abc/blob/main/LeanPolyABC/Corollaries/FltCatalan.lean).

Related: #16060

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+open scoped Polynomial Classical
+
+open Polynomial UniqueFactorizationMonoid
+
+variable {k : Type*} [Field k]
+
+
+/-- Auxiliary theorems for `max₃ a b c = max (max a b) c`. -/
+def max₃ (a b c : Nat) : Nat :=
```
Comment (jcpaik) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:40]: Do we want this auxiliary function for maximum of three natural numbers in the statement of Mason--Stothers? I wonder how other parts of mathlib represents maximum values of three numbers.

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 2]
```diff
@@ -0,0 +1,253 @@
+/-
+Copyright (c) 2024 Jineon Back and Seewoo Lee. All rights reserved.
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:2]:
```suggestion
Copyright (c) 2024 Jineon Baek and Seewoo Lee. All rights reserved.
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+Authors: Jineon Baek, Seewoo Lee
+-/
+import Init.Data.Nat.Lemmas
+import Mathlib.Algebra.EuclideanDomain.Defs
+import Mathlib.Algebra.Polynomial.RingDivision
+import Mathlib.Algebra.Ring.Regular
+import Mathlib.RingTheory.Polynomial.Wronskian
+import Mathlib.RingTheory.Radical
+import Mathlib.RingTheory.UniqueFactorizationDomain
+import Mathlib.Algebra.Polynomial.FieldDivision
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:13]: Looks like a bit too many imports. Could you please write `#min_imports` at the bottom of the file and copy here whatever it gives you?

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+This file states and proves Mason-Stothers theorem, which is a polynomial version of ABC conjecture.
+For a (pairwise) coprime polynomials `a, b, c` (over a field) with `a + b + c = 0`, we have
+`max {deg(a), deg(b), deg(c)} + 1 ≤ deg(rad(abc))` or `a' = b' = c' = 0`.
+
+Proof is based on this online note by Franz Lemmermeyer http://www.fen.bilkent.edu.tr/~franz/ag05/ag-02.pdf,
+which is essentially based on Noah Snyder's paper "An Alternative Proof of Mason's Theorem",
+but slightly different.
+
+## TODO
+Prove polynomial FLT using Mason-Stothers theorem.
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:27]:
```suggestion

Prove polynomial FLT using Mason-Stothers theorem.
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+-/
+
+noncomputable section
+
+open scoped Polynomial Classical
+
+open Polynomial UniqueFactorizationMonoid UniqueFactorizationDomain
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:34]: No need to both `open` and `open scoped`
```suggestion
open scoped Classical

open Polynomial UniqueFactorizationMonoid UniqueFactorizationDomain
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+theorem max₃_lt {a b c d : Nat} : max₃ a b c < d ↔ a < d ∧ b < d ∧ c < d := by
+  rw [max₃, Nat.max_lt, Nat.max_lt, and_assoc]
+
+theorem max₃_le {a b c d : Nat} : max₃ a b c ≤ d ↔ a ≤ d ∧ b ≤ d ∧ c ≤ d := by
+  rw [max₃, Nat.max_le, Nat.max_le, and_assoc]
+
+
+/-- For a given polynomial `a`, `divRadical a` is `a` divided by its radical `radical a`.-/
+def divRadical (a : k[X]) : k[X] := a / (radical a)
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:66]:
```suggestion
def divRadical (a : k[X]) : k[X] := a / radical a
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+
+theorem hMul_radical_divRadical (a : k[X]) : (radical a) * (divRadical a) = a := by
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:69]:
```suggestion
theorem hMul_radical_divRadical (a : k[X]) : radical a * divRadical a = a := by
```
and below

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+  rw [divRadical]
+  rw [← EuclideanDomain.mul_div_assoc]
+  ·refine mul_div_cancel_left₀ _ (radical_ne_zero a)
+  exact radical_dvd_self a
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:73]:
```suggestion
  rw [divRadical, ← EuclideanDomain.mul_div_assoc (radical_dvd_self a),
    mul_div_cancel_left₀ _ (radical_ne_zero a)]
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+  exact right_ne_zero_of_mul ha
+
+theorem divRadical_isUnit {u : k[X]} (hu : IsUnit u) : IsUnit (divRadical u) := by
+  rwa [divRadical, radical_unit_eq_one hu, EuclideanDomain.div_one]
+
+theorem eq_divRadical {a x : k[X]} (h : (radical a) * x = a) : x = divRadical a := by
+  apply EuclideanDomain.eq_div_of_mul_eq_left (radical_ne_zero a)
+  rwa [mul_comm]
+
+theorem divRadical_hMul {a b : k[X]} (hc : IsCoprime a b) :
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:86]:
```suggestion
theorem divRadical_mul {a b : k[X]} (hc : IsCoprime a b) :
```
and everywhere else

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+  cases' hdab with hda hdb
+  rw [wronskian]
+  rw [hda, hdb]; simp only [MulZeroClass.mul_zero, MulZeroClass.zero_mul, sub_self]
+
+
+protected theorem IsCoprime.divRadical {a b : k[X]} (h : IsCoprime a b) :
+    IsCoprime (divRadical a) (divRadical b) := by
+  rw [← hMul_radical_divRadical a] at h
+  rw [← hMul_radical_divRadical b] at h
+  exact h.of_mul_left_right.of_mul_right_right
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:179]: Could all of this move to a file under `Mathlib.RingTheory.Polynomial`? Either `Mathlib.RingTheory.Polynomial.Wronskian` or a new file `Mathlib.RingTheory.Polynomial.DivRadical`? And can you make that into a separate PR?

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+      _ = (a * b * c).natDegree := by
+        rw [mul_comm _ (radical _)]; rw [hMul_radical_divRadical (a * b * c)]
+      _ = a.natDegree + b.natDegree + c.natDegree := by
+        rw [Polynomial.natDegree_mul hab hc, Polynomial.natDegree_mul ha hb]
+  rw [t3] at t4
+  exact Nat.lt_of_add_lt_add_left t4
+
+private theorem rot3_add {a b c : k[X]} : a + b + c = b + c + a := by ring_nf
+
+private theorem rot3_mul {a b c : k[X]} : a * b * c = b * c * a := by ring_nf
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:209]: These already exist! They are called `add_rotate` and `mul_rotate`

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+
+/-- Polynomial ABC theorem. -/
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:211]:
```suggestion
/-- **Polynomial ABC theorem**. -/
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+    convert abcdr_dvd_w using 1
+    rw [← divRadical_hMul hab]
+    rw [← divRadical_hMul _]
+    exact hca.symm.mul_left hbc
+  by_cases hw : w = 0
+  · right
+    rw [hw] at wab wbc
+    cases' hab.wronskian_eq_zero_iff.mp wab.symm with ga gb
+    cases' hbc.wronskian_eq_zero_iff.mp wbc.symm with _ gc
+    refine ⟨ga, gb, gc⟩
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:244]:
```suggestion
    exact ⟨ga, gb, gc⟩
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+  · left
+    rw [max₃_add_distrib_right, max₃_le]
+    constructor
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:247]: Several `constructor`s can be turned into one `refine`
```suggestion
    refine ⟨?_, ?_, ?_⟩
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/NumberTheory/FLT/MasonStothers.lean:40]: We usually just write `max (max a b) c`, or `a ⊔ b ⊔ c` if you prefer infixes. But also you are using it to state `max (max a b) c ≤ d`, in which situation we much prefer writing `a ≤ d ∧ b ≤ d ∧ c ≤ d`

File: Mathlib/NumberTheory/FLT/MasonStothers.lean (hunk 10 above)
Comment (seewoo5) [reply Mathlib/NumberTheory/FLT/MasonStothers.lean:179]: I'll put under `Mathlib.RingTheory.Polynomial` then (IMO these theorems are to specific so I'm not sure if it worths to make new files for it).

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 15]
```diff
@@ -233,21 +225,20 @@ theorem Polynomial.abc {a b c : k[X]} (ha : a ≠ 0) (hb : b ≠ 0) (hc : c ≠
     have abdr_dvd_w := cop_ab_dr.mul_dvd adr_dvd_w bdr_dvd_w
     have abcdr_dvd_w := cop_abc_dr.mul_dvd abdr_dvd_w cdr_dvd_w
     convert abcdr_dvd_w using 1
-    rw [← divRadical_hMul hab]
-    rw [← divRadical_hMul _]
+    rw [← divRadical_mul hab]
+    rw [← divRadical_mul _]
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:229]:
```suggestion
    rw [← divRadical_mul]
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean (hunk 10 above)
Comment (seewoo5) [reply Mathlib/NumberTheory/FLT/MasonStothers.lean]: I just realized that first half of the theorems & definition of `divRadical` actually makes sense for `EuclideanDomain` as well, but not sure if I put these under the "euclidean domain" files. Do you have any suggestions? I think another way is to put `divRadical`-related theorems, up to `divRadical_dvd_self`, in `Mathlib.RingTheory.Radical` with `EuclideanDomain` version, and put the rest in `Mathlib.RingTheory.Polynomial`? (Honestly I've never seen any mathematical proofs using $x / \mathrm{rad}(x)$ except for Mason-Stothers...)

Comment (seewoo5) [reply Mathlib/NumberTheory/FLT/MasonStothers.lean]: For `dvd_derivative_iff` and `IsCoprime.wronskian_eq_zero_iff`, I think these are also seperate things. I'd put the second one under `Mathlib.RingTheory.Polynomial.Wronskian`, but this depends on the first theorem and I don't know where's the best place to put it - do you have any suggestions? (these two are not included in #18452 for now)

Comment (YaelDillies) [reply Mathlib/NumberTheory/FLT/MasonStothers.lean]: No sorry, no suggesyion. I would play around with the myself but I am momentarily incapactitated by apainful wrist

**Comment by YaelDillies** [comment]
Can you please merge maqter and reduce the diff?

File: Mathlib/NumberTheory/FLT/MasonStothers.lean (hunk 1 above)
Comment (seewoo5) [reply Mathlib/NumberTheory/FLT/MasonStothers.lean:34]: Should we remove all `max₃` related apis and work with `max`-related things then?

Comment (YaelDillies) [reply Mathlib/NumberTheory/FLT/MasonStothers.lean:34]: Yes that would great 🙏

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+Copyright (c) 2024 Jineon Baek and Seewoo Lee. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jineon Baek, Seewoo Lee
+-/
+import Mathlib.RingTheory.Polynomial.Radical
+
+/-!
+# Mason-Stothers theorem
+
+This file states and proves Mason-Stothers theorem, which is a polynomial version of ABC conjecture.
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:11]:
```suggestion
This file states and proves the Mason-Stothers theorem, which is a polynomial version of the ABC conjecture.
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+For a (pairwise) coprime polynomials `a, b, c` (over a field) with `a + b + c = 0`, we have
```
Comment (YaelDillies) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:12]:
```suggestion
For (pairwise) coprime polynomials `a, b, c` (over a field) with `a + b + c = 0`, we have
```

**Review by YaelDillies — APPROVED** [review]
maintainer merge

apart from that. Thanks!

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+      natDegree c + 1 ≤ (radical (a * b * c)).natDegree
+    ) ∨
+      derivative a = 0 ∧ derivative b = 0 ∧ derivative c = 0 := by
+  -- Utility assertions
+  have wbc := wronskian_eq_of_sum_zero hsum
+  set w := wronskian a b with wab
+  have wca : w = wronskian c a := by
+    rw [add_rotate] at hsum
+    have h := wronskian_eq_of_sum_zero hsum
+    rw [← wbc] at h; exact h
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:72]:
```suggestion
    simpa only [← wbc] using wronskian_eq_of_sum_zero hsum
```

File: Mathlib/NumberTheory/FLT/MasonStothers.lean [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+    have cop_ab_dr := hab.divRadical
+    have cop_bc_dr := hbc.divRadical
+    have cop_ca_dr := hca.divRadical
+    have cop_abc_dr := cop_ca_dr.symm.mul_left cop_bc_dr
+    have abdr_dvd_w := cop_ab_dr.mul_dvd adr_dvd_w bdr_dvd_w
+    have abcdr_dvd_w := cop_abc_dr.mul_dvd abdr_dvd_w cdr_dvd_w
+    convert abcdr_dvd_w using 1
+    rw [← divRadical_mul hab]
+    rw [← divRadical_mul]
+    exact hca.symm.mul_left hbc
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/FLT/MasonStothers.lean:88]:
```suggestion
  have abc_dr_dvd_w : divRadical (a * b * c) ∣ w := by
    have adr_dvd_w := divRadical_dvd_wronskian_left a b
    have bdr_dvd_w := divRadical_dvd_wronskian_right a b
    have cdr_dvd_w := divRadical_dvd_wronskian_right b c
    rw [← wab] at adr_dvd_w bdr_dvd_w
    rw [← wbc] at cdr_dvd_w
    convert (hca.divRadical.symm.mul_left hbc.divRadical).mul_dvd
      (hab.divRadical.mul_dvd adr_dvd_w bdr_dvd_w) cdr_dvd_w using 1
    rw [← divRadical_mul hab, ← divRadical_mul]
    exact hca.symm.mul_left hbc
```

**Review by riccardobrasca — COMMENTED** [review]
I find the proof easier to read without too many `have` statement. On the other hand, can you add a little bit of math explanation?

I am not delegating it immediately since I *suspect* there is a diamond somewhere, let me investigate this.

**Comment by riccardobrasca** [comment]
I pushed a small modification to another file. The lines
```lean
noncomputable section

open scoped Classical
```
can now be removed, adding `[DecidableEq k]`. I suggest a slightly modified proof of the main result:
```lean
/-- **Polynomial ABC theorem.** -/
theorem Polynomial.abc {a b c : k[X]} (ha : a ≠ 0) (hb : b ≠ 0) (hc : c ≠ 0) (hab : IsCoprime a b)
    (hbc : IsCoprime b c) (hca : IsCoprime c a) (hsum : a + b + c = 0) :
    ( natDegree a + 1 ≤ (radical (a * b * c)).natDegree ∧
      natDegree b + 1 ≤ (radical (a * b * c)).natDegree ∧
      natDegree c + 1 ≤ (radical (a * b * c)).natDegree ) ∨
      derivative a = 0 ∧ derivative b = 0 ∧ derivative c = 0 := by
  -- Utility assertions
  have wbc := wronskian_eq_of_sum_zero hsum
  set w := wronskian a b with wab
  have wca : w = wronskian c a := by
    rw [add_rotate] at hsum
    simpa only [← wbc] using wronskian_eq_of_sum_zero hsum
  have abc_dr_dvd_w : divRadical (a * b * c) ∣ w := by
    have adr_dvd_w := divRadical_dvd_wronskian_left a b
    have bdr_dvd_w := divRadical_dvd_wronskian_right a b
    have cdr_dvd_w := divRadical_dvd_wronskian_right b c
    rw [← wab] at adr_dvd_w bdr_dvd_w
    rw [← wbc] at cdr_dvd_w
    rw [divRadical_mul (hca.symm.mul_left hbc), divRadical_mul hab]
    exact (hca.divRadical.symm.mul_left hbc.divRadical).mul_dvd
      (hab.divRadical.mul_dvd adr_dvd_w bdr_dvd_w) cdr_dvd_w
  by_cases hw : w = 0
  · right
    rw [hw] at wab wbc
    cases' hab.wronskian_eq_zero_iff.mp wab.symm with ga gb
    cases' hbc.wronskian_eq_zero_iff.mp wbc.symm with _ gc
    exact ⟨ga, gb, gc⟩
  · left
    refine ⟨?_, ?_, ?_⟩
    · rw [mul_rotate] at abc_dr_dvd_w ⊢
      apply abc_subcall wbc <;> assumption
    · rw [← mul_rotate] at abc_dr_dvd_w ⊢
      apply abc_subcall wca <;> assumption
    · apply abc_subcall wab <;> assumption
```

**Comment by jcpaik** [comment]
> I am not delegating it immediately since I _suspect_ there is a diamond somewhere, let me investigate this.

I don't get what you mean by this. Are you saying that there is a diamond (I suppose the diamond problem of instance?) in this PR, or some other part of the mathlib that this PR depends on?

**Comment by riccardobrasca** [comment]
> > I am not delegating it immediately since I _suspect_ there is a diamond somewhere, let me investigate this.
> 
> I don't get what you mean by this. Are you saying that there is a diamond (I suppose the diamond problem of instance?) in this PR, or some other part of the mathlib that this PR depends on?

It's in another file, but it is fixed by my commit, you can forget about it.

**Review by riccardobrasca — COMMENTED** [review]
Thanks!

bors d+

**Comment by seewoo5** [comment]
Thank you all!

bors r+

**Pull request closed without merging**